### PR TITLE
device: fixed Ledger Nano S device selection on Windows/Mac

### PR DESCRIPTION
**Pull request by xiphon**
#4534

### CODE REVIEW

File: src/device/device_io_hid.hpp [hunk 1]
```diff
@@ -1,112 +1,109 @@
...
+     * |----------------------------------------------------------|
+     */
+
+
+    struct hid_conn_params {
+      unsigned int vid; 
+      unsigned int pid; 
+      int interface_number;
+      unsigned short usage_page;
    };
```
Comment (ph4r05) [inline src/device/device_io_hid.hpp]: `^M` ?

Comment (moneromooo-monero) [reply src/device/device_io_hid.hpp]: It's DOS style newlines (\r), I reckon an editor took it upon itself to mangle the whole file.

File: src/device/device_io_hid.hpp [hunk 2]
```diff
@@ -1,112 +1,109 @@
...
+      void connect(unsigned int vid, unsigned  int pid, boost::optional<int> interface_number, boost::optional<unsigned short> usage_page);
+      bool connected() const;
+      int  exchange(unsigned char *command, unsigned int cmd_len, unsigned char *response, unsigned int max_resp_len);
+      void disconnect();
+      void release();
+    };
+  };
+};
+
+#endif //#if defined(HAVE_HIDAPI)
```
[the remark this answers is not in the document]
Comment (xiphon) [reply src/device/device_io_hid.hpp]: Looks like the way to compile binaries without Ledger support. Dunno if this actually works, didn't test it, but there are some related defines in CMakeLists and other headers.

File: src/device/device_io_hid.cpp [hunk 3]
```diff
@@ -69,29 +70,53 @@ namespace hw {
...
+      LOG_PRINT_L2( "Looking for HID Device with" <<
+                    " interface_number " << (interface_number ? std::to_string(interface_number.value()) : "ANY") <<
+                    ((interface_number && usage_page) ? " or" : " and") <<
+                    " usage_page " << (usage_page ? std::to_string(usage_page.value()) : "ANY"));
+
+      hid_device_info *result = nullptr;
+      bool select_any = !interface_number && !usage_page;
+      for (; devices_list != nullptr; devices_list = devices_list->next) {
+        BOOST_SCOPE_EXIT(&devices_list, &result) {
+          LOG_PRINT_L2( (result == devices_list ? "SELECTED" : "SKIPPED ") <<
```
[the remark this answers is not in the document]
Comment (xiphon) [reply src/device/device_io_hid.cpp]: Using this space to indent the log messages, take a look at:
```
DEBUG   device.io       src/device/device_io_hid.cpp:80 Looking for HID Device with interface_number 0 or usage_page 65440
DEBUG   device.io       src/device/device_io_hid.cpp:90 SKIPPED  HID Device path \\?\hid#vid_2c97&pid_0001&mi_01#7&3478eb0e&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030} interface_number 1 usage_page 61904
DEBUG   device.io       src/device/device_io_hid.cpp:90 SELECTED HID Device path \\?\hid#vid_2c97&pid_0001&mi_00#7&307a4322&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030} interface_number 0 usage_page 65440
```

File: src/device/device_io_hid.cpp [hunk 4]
```diff
@@ -69,29 +70,53 @@ namespace hw {
...
+          continue;
+        }
+
+        if (select_any) {
+          result = devices_list;
+        } else if (interface_number && devices_list->interface_number == interface_number.value()) {
+          result = devices_list;
+        } else if (usage_page && devices_list->usage_page == usage_page.value()) {
+          result = devices_list;
+        }
```
[the remark this answers is not in the document]
Comment (xiphon) [reply src/device/device_io_hid.cpp]: Updated the message:

interface_number | usage_page |  log
--- | --- | ---
not set | not set | `any HID Device`
not set | set | `HID Device with usage_page X`
set | not set | `HID Device with interface_number X`
set | set | `HID Device with interface_number X or usage_page X`

**Comment by selsta** [comment]
I don’t know if this fixes the issues as I can’t reproduce them anymore, but I can confirm that this dosen’t break anything on macOS.

**Pull request merged**